ecommerce/extensions/payment/processors/xstack.py

```diff
@@ -0,0 +1,148 @@
+""" PostEx payment processing. """
+from __future__ import absolute_import, unicode_literals
+
+import logging
+from collections import OrderedDict
+from urllib.parse import urlencode
+
+from django.urls import reverse
+from oscar.apps.payment.exceptions import GatewayError
+from ecommerce.extensions.payment.processors import BasePaymentProcessor, HandledProcessorResponse
+
+logger = logging.getLogger(__name__)
+
+
+class XStack(BasePaymentProcessor):
+    """XStack API implementation."""
+
+    NAME = 'xstack'
+
+    def __init__(self, site):
+        """
+        Constructs a new instance of the XStack payment processor.
+
+        Raises:
+            KeyError: If a required setting is not configured for this payment processor
+        """
+        super(XStack, self).__init__(site)
+
+    @staticmethod
+    def get_courseid_title(line):
+        """
+        Get CourseID & Title from basket item
+
+        Arguments:
+            line: basket item
+
+        Returns:
+             Concatenated string containing course id & title if exists.
+        """
+        courseid = ''
+        line_course = line.product.course
+        if line_course:
+            courseid = "{}|".format(line_course.id)
+        return courseid + line.product.title
+
+    @staticmethod
+    def create_ordered_dict(data):
+        """Create ordered dict from list of tuples."""
+        ordered_data = OrderedDict()
+        for key, value in data:
+            ordered_data[key] = value
+        return ordered_data
+
+    def get_transaction_parameters(self, basket, request=None, use_client_side_checkout=False, **kwargs):
+        """
+        Create a new XStack Intent api url.
+
+        Arguments:
+            basket (Basket): The basket of products being purchased.
+            request (Request, optional): A Request object which is used to construct PayPal's `return_url`.
+            use_client_side_checkout (bool, optional): This value is not used.
+            **kwargs: Additional parameters; not used by this method.
+
+        Returns:
+            dict: XStack-specific parameters required to complete a transaction. Must contain a URL
+                to which users can be directed in order to approve a newly created payment.
+        """
+        logger.info('Starting xstack payment url creation')
+        order_id = basket.order_number
+        amount = basket.total_incl_tax
+        user_name = basket.owner.username
+        data = self.create_ordered_dict([
+            ('customerName', user_name),
+            ('amount', amount),
+            ('orderRefNum', order_id),
+        ])
+        logger.info('Generated data: {}'.format(data))
+        query_str = urlencode(data)
+        logger.info('Generated str: {}'.format(query_str))
```

Comment:
Too many logs, add a TODO comment here to remove these once everything has been finalised